jgrapht-core/src/main/java/org/jgrapht/util/DoublyLinkedList.java

```diff
@@ -0,0 +1,485 @@
+/*
+ * (C) Copyright 2018-2018, by Timofey Chudakov and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.util;
+
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * This is the implementation of the doubly linked list data structure, which gives access to the internal
+ * list nodes the data is stored in. The primary goal of this is to be able to delete an element from this
+ * list in $\mathcal{O}(1)$ via the list node it is stored in.
+ * <p>
+ * The iterators over this list have a <i>fail-fast</i> behaviour meaning that they throw a
+ * {@link ConcurrentModificationException} after they detect a structural modification of the list,
+ * that they're not responsible for.
+ *
+ * @param <E> the list element type
+ * @author Timofey Chudakov
+ */
+public class DoublyLinkedList<E> implements Iterable<E> {
```

Comment:
For a future PR, it might be good to see whether we can also implement the `List` interface.